File: Binary Indexed Tree.cpp
#include<bits/stdc++.h>
using namespace std;
#define ll long long
ll lsb(ll idx)
{
    return idx & (-idx);
}
void update(ll *bit,ll node_idx,ll n,ll val)
{
    // bottom to top
    while(node_idx<=n)
    {
        bit[node_idx]+=val;
        // go to parent
        node_idx+=lsb(node_idx);
    }
}
ll get_sum(ll node_idx,ll *bit)
{
    ll sum=0;
    // Top to bottom
    while(node_idx>=1)
    {
        sum+=bit[node_idx];
        // go to child
        node_idx-=lsb(node_idx);
    }
    return sum;
}
ll *create_bit(ll *a,ll n)
{
    ll *bit=new ll[n+1];
    for(ll i=1;i<=n;i++)
        bit[i]=0;
    for(ll i=1;i<=n;i++)
        update(bit,i,n,a[i]);
    return bit;
}
int main()
{
    ll n,lb,ub,update_val,index;
    cout<<"Enter the size : ";
    cin>>n;
    ll a[n+1];
    cout<<"Enter the element : ";
    for(ll i=1;i<=n;i++)
        cin>>a[i];
    ll *bit=create_bit(a,n);
    cout<<"Enter the range(a,b) : ";
    cin>>lb>>ub;
    cout<<get_sum(ub,bit)-get_sum(lb-1,bit)<<endl;
    cout<<"Enter the index and update value : ";
    cin>>index>>update_val;
    update(bit,index,n,update_val);
    cout<<"After update"<<endl;
    cout<<"Enter the range(a,b) : ";
    cin>>lb>>ub;
    cout<<get_sum(ub,bit)-get_sum(lb-1,bit)<<endl;
}

File: Binary Indexed Tree.py
def lsb(idx):
    return idx & (-idx)
def update(bit,node_idx,n,val):
    while(node_idx<=n):
        bit[node_idx]+=val
        # go to parent or bottom to top
        node_idx+=lsb(node_idx)
def get_sum(node_idx,bit):
    sum=0
    while node_idx>=1:
        sum+=bit[node_idx]
        node_idx-=lsb(node_idx)
    return sum
def create_bit(a,n):
    bit=[0]*(n+1)
    for i in range(1,n+1):
        update(bit,i,n,a[i])
    return bit
if __name__ == "__main__" :
    a=[0,1,2,3,4,5]
    n=len(a)-1
    bit=create_bit(a,n)
    upper_bit=4
    lower_bit=2
    print(get_sum(upper_bit,bit)-get_sum(lower_bit-1,bit))
    update(bit,1,n,9) # tree,index_no,size,value
    print(get_sum(upper_bit,bit)-get_sum(lower_bit-1,bit))
